lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs

```diff
@@ -0,0 +1,1015 @@
+            let coinSelected = TokenBundle.coin $
+                    addExtraSource extraSource balanceSelected
+            monitor $ counterexample $ unlines
+                [ "balance selected: " <> show balanceSelected
+                ]
+            assert $ coinSelected >= TokenBundle.coin balanceRequested
+            assert $ balanceSelected /= TokenBundle.empty
+  where
+    almostEverything = TokenBundle.subtract
+        (view #balance index)
+        (TokenBundle.fromCoin (Coin 1))
+
+prop_runSelection_UTxO_moreThanEnough
+    :: Maybe Coin
+    -> Small UTxOIndex
+    -> Property
+prop_runSelection_UTxO_moreThanEnough extraSource (Small index) = monadicIO $ do
+    SelectionState {selected, leftover} <-
+        run $ runSelection extraSource index balanceRequested
+    let balanceSelected = view #balance selected
+    let balanceLeftover = view #balance leftover
+    monitor $ cover 80
+        (assetsRequested `Set.isProperSubsetOf` assetsAvailable)
+        "assetsRequested ⊂ assetsAvailable"
+    monitor $ cover 50 (Set.size assetsRequested >= 4)
+        "size assetsRequested >= 4"
+    monitor $ counterexample $ unlines
+        [ "balance available: " <> show balanceAvailable
+        , "balance requested: " <> show balanceRequested
+        , "balance selected:  " <> show balanceSelected
+        , "balance leftover:  " <> show balanceLeftover
+        ]
+    assert $ balanceRequested `leq` addExtraSource extraSource balanceSelected
+    assert $ balanceAvailable == balanceSelected <> balanceLeftover
+  where
+    assetsAvailable = TokenBundle.getAssets balanceAvailable
+    assetsRequested = TokenBundle.getAssets balanceRequested
+    balanceAvailable = view #balance index
+    balanceRequested = adjustAllQuantities (`div` 8) $
+        cutAssetSetSizeInHalf balanceAvailable
+
+prop_runSelection_UTxO_muchMoreThanEnough
+    :: Maybe Coin
+    -> Blind (Large UTxOIndex)
+    -> Property
+prop_runSelection_UTxO_muchMoreThanEnough extraSource (Blind (Large index)) =
+    -- Generation of large UTxO sets takes longer, so limit the number of runs:
+    withMaxSuccess 100 $
+    checkCoverage $
+    monadicIO $ do
+        SelectionState {selected, leftover} <-
+            run $ runSelection extraSource index balanceRequested
+        let balanceSelected = view #balance selected
+        let balanceLeftover = view #balance leftover
+        monitor $ cover 80
+            (assetsRequested `Set.isProperSubsetOf` assetsAvailable)
+            "assetsRequested ⊂ assetsAvailable"
+        monitor $ cover 50 (Set.size assetsRequested >= 4)
+            "size assetsRequested >= 4"
+        monitor $ counterexample $ unlines
+            [ "balance available: " <> show balanceAvailable
+            , "balance requested: " <> show balanceRequested
+            , "balance selected:  " <> show balanceSelected
+            , "balance leftover:  " <> show balanceLeftover
+            ]
```

Comment:
LIkewise, we can use `pretty . TokenBundle.Nested` here to make the output of `counterexample` more readable.